rust/src/main.rs

```diff
@@ -16,25 +16,25 @@ struct BenchHandler {
 
 impl ws::Handler for BenchHandler {
     fn on_message(&mut self, msg: ws::Message) -> ws::Result<()> {
-        if let Some((Some(msg_type), payload)) = msg.into_text().ok()
-            .map(|v| v.as_str().to_owned())
-            .and_then(|body: String| serde_json::from_str(body.as_str()).ok())
-            .and_then(|j: Value| j.as_object().map(move |obj: &Map<String, Value>| {
-                let t = obj.get("type").and_then(|t| t.as_str()).map(|s| s.to_owned());
-                let p: Value = obj.get("payload").unwrap_or(&Value::Null).clone();
-                (t, p)
-            })) {
+        if let Ok(body) = msg.into_text() {
+            if let Some((Some(msg_type), payload)) = serde_json::from_str(body.as_str()).ok()
+                .and_then(|j: Value| j.as_object().map(move |obj: &Map<String, Value>| {
+                    let t = obj.get("type").and_then(|t| t.as_str()).map(|s| s.to_owned());
```

Comment:
I'll bet `t` (and thus `msg_type`) doesn't have to be owned, considering you only call `as_ref` on it later.  
I see you've eliminated other similar allocations though, so that's good.